Add kraken OCR engine

stweil opened the pull request.

stweil committed:
Add OCR models Fraktur and Latin for Tesseract

Both are not language specific, but support historic and current scripts
used by many European languages.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add new OCR engine kraken

Kraken is an Open Source OCR engine with trainable segmentation and
OCR models. It can work with printed and handwritten texts.

This initial implementation comes with two generic OCR models which
can be used on a wide range of German publications, but also with
other languages which are based on Latin script.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
WebProfilerBundle

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Remove unneeded code for Transkribus OCR engine

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add script for kraken OCR

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Update KrakenEngine to support language selection

Cropping is also implemented now, but still untested.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add models for kraken OCR

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil commented:
This is a first implementation to support the Open Source kraken OCR engine.

I created this draft pull request to allow public review and comments although the implementation is still incomplete. OCR for cropped image still needs testing, and there is also currently no unit test code for the new engine.

stweil committed:
Suppress warning from phpcs because usage of popen

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add austriannewspapers model for kraken

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add missing documentation for new OCR engine kraken (required for CI test)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

samwilson commented:
Also, would you mind opening a Phabricator task for this work, so it can be tracked there? Thanks!

stweil commented:
> Also, would you mind opening a Phabricator task for this work, so it can be tracked there? Thanks!

Done, see https://phabricator.wikimedia.org/T345055. I also updated the PR here to solve a merge conflict.

stweil commented:
See also my [test installation](https://kraken-ocr.wmcloud.org/).

stweil committed:
Update package.json

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Update package-lock.json

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Support segmentation model for kraken OCR engine

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Update API version for new release with kraken OCR engine

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add new API /api/available_segmentation_models

Segmentation models are currently only supported for kraken.
All other OCR engines return an empty list.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add segmentation model for kraken OCR

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add more OCR models for Tesseract

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
npm: Add missing dependency stylelint

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Fix iteration over Transkribus line models

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Fix description of OpenAPI parameters langs and crop

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add Transkribus model german-fraktur-19th-20th-century

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Fix kraken_ocr script

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add new OCR parameter to normalize the result text

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil commented:
The test installation is meanwhile available on https://kraken-ocr.wmcloud.org/.

stweil committed:
Fix typo in name of newly introduced method (instatiate -> instantiate)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Modify title shown on test web page

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Fix code injection

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Improve code to fix code injection

Signed-off-by: Stefan Weil <sw@weilnetz.de>

Parthiv-M commented on src/Controller/OcrController.php:
```diff
@@ -290,6 +334,28 @@ public function apiAvailableLangsAction(): JsonResponse {
+	/**
+	 * Get a list of available segmentation models for use with a specific OCR engine.
+	 *
+	 * @Route("/api/available_segmentation_models", name="apiSegmentationModels", methods={"GET"})
```
This route, since it is related to Kraken, should be named `/api/kraken/available_segmentation_models`

stweil replied:
Having segmentation models is not kraken specific. All OCR processes require a segmentation step, and if that step uses AI, it also requires a model. That's why I did not use a route with "kraken" here. So even if it is currently only used for kraken, I'd suggest to use a generic route.

Parthiv-M commented on public/langs.json:
```diff
@@ -265,6 +286,11 @@
             "htr": 45902
         }
     },
+    "german-fraktur-19th-20th-century": {
+        "transkribus": {
+            "htr": 37738
+        }
+    },
```
This Transkribus model should not be added along with Kraken changes, it would be better to separate it out

Parthiv-M commented on src/Engine/TranskribusEngine.php:
```diff
@@ -118,9 +118,6 @@ public function getResult(
-
 		$points = '';
 		if ( $crop ) {
 			$x = $crop['x'];
```
We'd prefer to isolate changes other than those related to Kraken to another PR!

Parthiv-M commented on tests/Engine/EngineBaseTest.php:
```diff
@@ -28,11 +32,13 @@ class EngineBaseTest extends OcrTestCase {
 
-		$this->tesseractEngine = $this->instatiateEngine( 'tesseract' );
+		$this->transkribusEngine = $this->instantiateEngine( 'transkribus' );
 
-		$this->transkribusEngine = $this->instatiateEngine( 'transkribus' );
+		$this->krakenEngine = $this->instantiateEngine( 'kraken' );
```
`instantiate()` fixed in #115

Parthiv-M commented on config/packages/nelmio_api_doc.yaml:
```diff
@@ -2,10 +2,11 @@ nelmio_api_doc:
-            description: A web service for Tesseract, Google and Transkribus OCR engines.
-            version: 1.0.0
+            description: A web service for Kraken, Tesseract, Google and Transkribus OCR engines.
+            version: 1.4.0
```
I believe it has been bumped up to 1.4.4 now.

Parthiv-M commented on config/packages/nelmio_api_doc.yaml:
```diff
@@ -2,10 +2,11 @@ nelmio_api_doc:
         path_patterns:
             - ^/api$
             - ^/api/available_langs$
+            - ^/api/available_segmentation_models$
```
Will need to change this in accordance with my comment on route path

Parthiv-M commented on i18n/en.json:
```diff
@@ -1,6 +1,6 @@
 {
     "@metadata": {},
-    "title": "WikimediaOCR",
+    "title": "WikimediaOCR – Kraken Test",
```
This should remain as **WikimediaOCR**

Parthiv-M commented on package.json:
```diff
@@ -13,6 +13,7 @@
         "regenerator-runtime": "^0.13.11",
         "select2": "^4.0.13",
         "select2-bootstrap-theme": "0.1.0-beta.10",
+        "stylelint": "^15.10.3",
```
`stylelint` can be removed from this PR as well

Parthiv-M commented on package.json:
```diff
@@ -23,5 +24,6 @@
         "test": "grunt test"
-    }
+    },
+    "dependencies": {}
```
Empty entries should be removed

Parthiv-M commented on src/Engine/EngineBase.php:
```diff
@@ -65,6 +72,7 @@ abstract class EngineBase {
 		'fro' => 'Franceis, François, Romanz (1400-1600)',
 		'ger-hd-m1' => 'Transkribus German handwriting M1',
 		'ger-15' => '15th-16th century German',
+		'german-fraktur-19th-20th-century' => 'German Fraktur 19th-20th century',
```
This is a Transkribus model and needs to be removed from this PR

Parthiv-M reviewed (changes requested):
This review focuses on the following main things
- Files related to Transkribus have also been modified, which I think is not ideal
- A new Transkribus model has also been added, which should be removed
- Normalise text has been added here for all engines (we would want that in a separate PR)
- Renaming the newly added API route

Overall, we'd like to keep the non-Kraken changes out of the way before testing kraken once again

wrznr commented:
It is actually a pitty that this PR never got merged. It would have been a great tool for the transcription of texts on wikisource.

stweil commented:
So let me fix the conflicts first.

samwilson commented:
I've moved this PR to https://gitlab.wikimedia.org/toolforge-repos/ocr/-/merge_requests/2

stweil commented:
@samwilson, do I need a new account for gitlab.wikimedia.org? Or is one of my existing accounts also valid there?

samwilson commented:
You should be able to log in there with your Wikimedia developer account. Let me know if you have any issues though.

I can bring the patch up to date. I feel bad that I dropped the ball on this last year. I'm also thinking that we need to do some Apache proxy config so that kraken requests get forwarded to the right server.

stweil commented:
I get this message on gitlab.wikimedia.org: "Your account is pending approval from your GitLab administrator and hence blocked. Please contact your GitLab administrator if you think this is an error."

samwilson commented:
I've added you to trusted-contributors: https://phabricator.wikimedia.org/feed/7635354101851466613/

I think you might need to log in to Phabricator with your developer account, at the moment you don't have a ldap username linked: https://phabricator.wikimedia.org/p/sweil/
